File: processing/syntactic/expressions/while_expr.py
from __future__ import annotations

from processing.codegen.code_builder import CodeBuilder
from processing.lexical.tokens.token_type import TokenType
from processing.syntactic.expressions.block_expr import BlockExpr, BlockType
from processing.syntactic.expressions.expr import Expr, child_property
from processing.syntactic.expressions.groups import StatementExpression
from processing.syntactic.parsing import parse_infix


class WhileExpr(StatementExpression):
    """ while_expr:
        'while' infix_expr block
        ['nobreak' block];
    """

    @child_property
    def condition(self) -> Expr:
        pass

    @child_property
    def block(self) -> BlockExpr:
        pass

    @child_property
    def no_break_block(self) -> BlockExpr:
        pass

    def __init__(
            self,
            parent: Expr = None,
            condition: Expr = None,
            block: BlockExpr = None,
            no_break_block: BlockExpr = None
    ):
        super().__init__(parent)
        self.condition = condition
        self.block = block
        self.no_break_block = no_break_block

    def parse(self) -> WhileExpr:
        self.stream.eat(TokenType.keyword_while)
        self.condition = parse_infix(self)
        self.block = BlockExpr(self).parse(BlockType.default)
        if self.stream.maybe_eat(TokenType.keyword_nobreak):
            self.no_break_block = BlockExpr(self).parse(BlockType.default)
        return self

    def to_axion(self, c: CodeBuilder):
        c += 'while ', self.condition, self.block
        if self.no_break_block:
            c += 'nobreak', self.no_break_block

    def to_csharp(self, c: CodeBuilder):
        c += 'while ', self.condition, self.block
        if self.no_break_block:
            raise NotImplementedError

    def to_python(self, c: CodeBuilder):
        c += 'while ', self.condition, self.block
        if self.no_break_block:
            c += 'else', self.no_break_block

File: Axion.Core/Processing/Syntactic/Expressions/WhileExpr.cs
using Axion.Core.Processing.CodeGen;
using static Axion.Core.Processing.Lexical.Tokens.TokenType;

namespace Axion.Core.Processing.Syntactic.Expressions {
    /// <summary>
    ///     <c>
    ///         while_expr:
    ///             'while' infix block
    ///             ['nobreak' block];
    ///     </c>
    /// </summary>
    public class WhileExpr : Expr, IStatementExpr, IDecoratedExpr {
        private Expr condition;

        public Expr Condition {
            get => condition;
            set => SetNode(ref condition, value);
        }

        private BlockExpr block;

        public BlockExpr Block {
            get => block;
            set => SetNode(ref block, value);
        }

        private BlockExpr noBreakBlock;

        public BlockExpr NoBreakBlock {
            get => noBreakBlock;
            set => SetNode(ref noBreakBlock, value);
        }

        public WhileExpr(
            Expr      parent,
            Expr      condition    = null,
            BlockExpr block        = null,
            BlockExpr noBreakBlock = null
        ) : base(parent) {
            Condition    = condition;
            Block        = block;
            NoBreakBlock = noBreakBlock;
        }

        public Expr Parse() {
            SetSpan(() => {
                Stream.Eat(KeywordWhile);
                Condition = InfixExpr.Parse(this);
                Block     = new BlockExpr(this).Parse();
                if (Stream.MaybeEat(KeywordNoBreak)) {
                    NoBreakBlock = new BlockExpr(this).Parse();
                }
            });
            return this;
        }

        public override void ToAxion(CodeWriter c) {
            c.Write("while ", Condition, Block);
            if (NoBreakBlock != null) {
                c.Write("nobreak", NoBreakBlock);
            }
        }

        public override void ToCSharp(CodeWriter c) {
            c.Write("while (", Condition);
            c.WriteLine(")");
            c.Write(Block);
        }

        public override void ToPython(CodeWriter c) {
            c.Write("while ", Condition, Block);
        }

        public override void ToPascal(CodeWriter c) {
            c.Write("while ", Condition, " do", Block);
        }
    }
}